File: Leetcode/621/solution.java
// https://leetcode.com/problems/task-scheduler/
class Solution {
    // Greedy Choice: schedule task that has most remaining instances
    // O(time) runtime, O(26) = O(1) space
    public int leastInterval(char[] tasks, int n) {
       int[] counts = new int[26];
       for (char c: tasks)
           counts[c - 'A']++;
       Arrays.sort(counts);

       int time = 0;
       while (counts[25] > 0) {
           int i = 0;
           while (i <= n) {
               if (counts[25] == 0)
                   break;
               if (i < 26 && counts[25 - i] > 0)
                   counts[25 - i]--;
               time++;
               i++;
           }
           Arrays.sort(counts);
       }
       return time;
   }
}

File: Leetcode/621/solution.py
# https://leetcode.com/problems/task-scheduler/
class Solution:
    # Greedy Choice: schedule task that has most remaining instances
    # O(time) runtime, O(26) = O(1) space
    def leastInterval(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        counts = [0] * 26
        for c in tasks:
            counts[ord(c) - ord('A')] += 1
        counts.sort()

        time = 0
        while counts[25] > 0:
            i = 0
            while i <= n:
                if counts[25] == 0:
                    break
                if i < 26 and counts[25-i] > 0:
                    counts[25 - i] -= 1;
                time += 1;
                i += 1;
            counts.sort()

        return time
